java/code/src/com/redhat/rhn/taskomatic/task/RecurringStateApplyJob.java

```diff
@@ -57,9 +60,14 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
     }
 
     private void scheduleAction(JobExecutionContext context, RecurringAction action) {
-        List<Long> minionIds = action.computeMinions().stream()
-                .map(m -> m.getId())
-                .collect(Collectors.toList());
+        Map<Boolean, List<MinionServer>> sepList = action.computeMinions().stream()
+              .collect(Collectors.partitioningBy(m -> m.checkIfInMaintenanceMode()));
+        sepList.get(true).stream()
+                       .forEach(m -> Log.info(m.getMinionId() + " is in Maintenance mode, skipping recurring action."));
```

Comment:
Here we need to be careful with the performance. When this one action has 10000 clients this gets slow.
While having `checkIfInMaintenanceMode()` for a single client is required, we should also have a function which does a mass checking.

Maybe something like this:

```
List<Long> systemIdsMaintenanceMode(List<MinionServer> minions)
```

Could be implemented in MaintenanceManager.java

Reply:
In the end, we check one-by-one, unfortunately (we don't have a method for checking whether a bunch of minions has maintenance windows in their calendars (currently it's implemented one by one)). But I agree that having the method suggested by you (`systemIdsMaintenanceMode(mids)`) should be there. Then we can optimize it somehow.

Reply:
One more note about the logging: In order to prevent a flood of log messages, I think the following would make sense:
- log a warning with the count of skipped minions
- log a debug message (in case log level is `>= DEBUG`) with the minion ids (can explode)

Reply:
What about mapping clients to maintenance schedules first and than just make this test for every maintenance schedule. This would limit the number of test a lot .

Reply:
that was my initial thinking was making a bool status thing on the server...